src/hotspot/share/opto/compile.cpp

```diff
@@ -4948,6 +4949,38 @@ bool Compile::coarsened_locks_consistent() {
   return true;
 }
 
+// Mark locking regions (identified by BoxLockNode) as unbalanced if
+// locks coarsening optimization removed Lock/Unlock nodes from them.
+// Such regions become unbalanced because coarsening only removes part
+// of Lock/Unlock nodes in region. As result we can't execute other
+// locks elimination optimizations which assume all code paths have
+// corresponding pair of Lock/Unlock nodes - they are balanced.
+void Compile::mark_coarsened_boxes() {
+  int count = coarsened_count();
+  for (int i = 0; i < count; i++) {
+    Node_List* locks_list = _coarsened_locks.at(i);
+    uint size = locks_list->size();
+    if (size > 0) {
+      AbstractLockNode* alock = locks_list->at(0)->as_AbstractLock();
+      BoxLockNode* box = alock->box_node()->as_BoxLock();
+      if (alock->is_coarsened()) {
+        // coarsened_locks_consistent(), which is called before this method, verifies
+        // that the rest of Lock/Unlock nodes on locks_list are also coarsened.
+        assert(!box->is_eliminated(), "regions with coarsened locks should not be marked as eliminated");
+        for (uint j = 1; j < size; j++) {
+          assert(locks_list->at(j)->as_AbstractLock()->is_coarsened(), "only coarsened locks are expected here");
+          BoxLockNode* this_box = locks_list->at(j)->as_AbstractLock()->box_node()->as_BoxLock();
+          assert(!this_box->is_eliminated(), "regions with coarsened locks should not be marked as eliminated");
+          if (box != this_box) {
+            box->set_unbalanced();
+            this_box->set_unbalanced();
+          }
+        }
+      }
+    }
+  }
+}
+
```

Comment:
Doesn't this end up marking even the simplest lock coarsening cases as unbalanced?
Is there a reason the above check can't be done as part of coarsened_locks_consistent()?

Reply:
> Doesn't this end up marking even the simplest lock coarsening cases as unbalanced?

Yes, the simplest lock coarsening makes locking regions as unbalanced.
See second code example in PR's description. Coarsening eliminates Unlock in regions 1 and Lock in region 2. Both regions become unbalanced.

> Is there a reason the above check can't be done as part of coarsened_locks_consistent()?

`Unbalanced` in coarsened_locks_consistent() is different from `Unbalanced` in `mark_coarsened_boxes()`
Or do you mean to recompile without lock coarsening when regions are unbalanced? But that will disable all coarsening, even valid one (for example, when no outer locking in my examples).

Reply:
> > Is there a reason the above check can't be done as part of coarsened_locks_consistent()?
> 
> `Unbalanced` in coarsened_locks_consistent() is different from `Unbalanced` in `mark_coarsened_boxes()` Or do you mean to recompile without lock coarsening when regions are unbalanced? But that will disable all coarsening, even valid one (for example, when no outer locking in my examples).

If we are simply making sure that all nodes in a group have the same BoxLock, then that check could be done earlier, in coarsened_locks_consistent(), or even as the nodes are added to the list.  But I guess it's possible to get a premature answer is nodes are removed from the list.

Reply:
I don't want to mix these two methods. They do different things:
 - `coarsened_locks_consistent()` checks that graph is correct otherwise bailout compilation
 - `mark_coarsened_boxes()` marks BoxLock nodes to prevent further locks optimizations

It could be done but it will complicate code in `coarsened_locks_consistent()` and you would still need 2 iterations:
 - first, you need to check that all `alock` on `locks_list` are marked as `coarsened` and also set flag if they reference different Boxes
 - only then you do second round over them to mark all referenced Box as unbalanced

And if we bailout from compilation Boxes marking will be useless work.